File: exercises/s02-python/src/s02_python/services/azure_openai_wrapper.py
import os
from typing import Any

from openai import AzureOpenAI

from .openai_wrapper import JsonSchemaFormat, extract_response_text

DEFAULT_API_VERSION = "2025-03-01-preview"


class AzureOpenAiWrapper:
    def __init__(
        self,
        api_key: str,
        azure_endpoint: str,
        deployment: str,
        api_version: str = DEFAULT_API_VERSION,
    ) -> None:
        self.deployment = deployment
        self._client = AzureOpenAI(
            api_key=api_key,
            azure_endpoint=azure_endpoint.rstrip("/"),
            api_version=api_version,
        )

    @classmethod
    def from_env(cls, deployment: str | None = None) -> "AzureOpenAiWrapper":
        api_key = os.environ.get("OPENAI_AZURE_API_KEY", "").strip()
        endpoint = os.environ.get("OPENAI_AZURE_ENDPOINT", "").strip()
        api_version = os.environ.get(
            "OPENAI_AZURE_API_VERSION", DEFAULT_API_VERSION
        ).strip()
        deploy = (
            deployment
            or os.environ.get("ELECTRICITY_AZURE_DEPLOYMENT", "").strip()
            or os.environ.get("OPENAI_AZURE_DEPLOYMENT", "").strip()
        )

        if not api_key:
            raise ValueError("Missing OPENAI_AZURE_API_KEY in environment")
        if not endpoint:
            raise ValueError("Missing OPENAI_AZURE_ENDPOINT in environment")
        if not deploy:
            raise ValueError(
                "Missing Azure deployment name (ELECTRICITY_AZURE_DEPLOYMENT or OPENAI_AZURE_DEPLOYMENT)"
            )

        return cls(api_key, endpoint, deploy, api_version)

    def responses_raw(self, payload: dict[str, Any]) -> dict[str, Any]:
        body = dict(payload)
        body.pop("model", None)
        response = self._client.responses.create(model=self.deployment, **body)
        return response.model_dump()

    def completion(
        self,
        model: str,
        system_prompt: str,
        user_prompt: str,
        json_schema_format: JsonSchemaFormat | None = None,
    ) -> str:
        del model
        payload = self._build_payload(system_prompt, user_prompt, json_schema_format)
        response = self._client.responses.create(
            model=self.deployment,
            **payload,
        )
        parsed = response.model_dump()
        text = extract_response_text(parsed)
        if text is None:
            raise RuntimeError(f"No text found in Azure response: {parsed}")
        return text

    @staticmethod
    def _build_payload(
        system_prompt: str,
        user_prompt: str,
        json_schema_format: JsonSchemaFormat | None,
    ) -> dict[str, Any]:
        payload: dict[str, Any] = {
            "input": [
                {
                    "role": "system",
                    "content": [{"type": "input_text", "text": system_prompt}],
                },
                {
                    "role": "user",
                    "content": [{"type": "input_text", "text": user_prompt}],
                },
            ],
        }

        if json_schema_format is not None:
            payload["text"] = {
                "format": {
                    "type": "json_schema",
                    "name": json_schema_format.name,
                    "strict": json_schema_format.strict,
                    "schema": json_schema_format.schema,
                }
            }

        return payload

File: exercises/s03-js/src/services/azure-openai-wrapper.js
import { extractResponseText } from "./openai-wrapper.js";

const DEFAULT_API_VERSION = "2025-03-01-preview";

const buildAzureResponsesUrl = (azureEndpoint, deployment, apiVersion) => {
  const normalizedEndpoint = azureEndpoint.replace(/\/+$/, "");
  const encodedDeployment = encodeURIComponent(deployment);
  const encodedApiVersion = encodeURIComponent(apiVersion);
  return `${normalizedEndpoint}/openai/deployments/${encodedDeployment}/responses?api-version=${encodedApiVersion}`;
};

const buildJsonSchemaFormat = (jsonSchemaFormat) => ({
  format: {
    type: "json_schema",
    name: jsonSchemaFormat.name,
    strict: jsonSchemaFormat.strict,
    schema: jsonSchemaFormat.schema
  }
});

const postJson = async (url, payload, headers) => {
  const response = await fetch(url, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      ...headers
    },
    body: JSON.stringify(payload)
  });

  const text = await response.text();
  let data;
  try {
    data = JSON.parse(text);
  } catch {
    throw new Error(`Response was not valid JSON: ${text}`);
  }

  const errorMessage =
    data && typeof data === "object" && !Array.isArray(data) && data.error?.message
      ? data.error.message
      : null;

  if (!response.ok) {
    throw new Error(`Request failed (${response.status}): ${errorMessage ?? "Unknown API error"}`);
  }

  if (errorMessage) {
    throw new Error(errorMessage);
  }

  return data;
};

export class AzureOpenAiWrapper {
  constructor(apiKey, azureEndpoint, deployment, apiVersion = DEFAULT_API_VERSION) {
    this.apiKey = apiKey;
    this.azureEndpoint = azureEndpoint;
    this.deployment = deployment;
    this.apiVersion = apiVersion;
  }

  static fromEnv(deployment = null) {
    const apiKey = process.env.OPENAI_AZURE_API_KEY?.trim() ?? "";
    const azureEndpoint = process.env.OPENAI_AZURE_ENDPOINT?.trim() ?? "";
    const apiVersion = process.env.OPENAI_AZURE_API_VERSION?.trim() ?? DEFAULT_API_VERSION;
    const deploy =
      deployment
      ?? process.env.ELECTRICITY_AZURE_DEPLOYMENT?.trim()
      ?? process.env.OPENAI_AZURE_DEPLOYMENT?.trim()
      ?? "";

    if (!apiKey) {
      throw new Error("Missing OPENAI_AZURE_API_KEY in environment");
    }

    if (!azureEndpoint) {
      throw new Error("Missing OPENAI_AZURE_ENDPOINT in environment");
    }

    if (!deploy) {
      throw new Error(
        "Missing Azure deployment name (ELECTRICITY_AZURE_DEPLOYMENT or OPENAI_AZURE_DEPLOYMENT)"
      );
    }

    return new AzureOpenAiWrapper(apiKey, azureEndpoint, deploy, apiVersion);
  }

  async responsesRaw(payload) {
    const body = { ...payload };
    delete body.model;

    const url = buildAzureResponsesUrl(this.azureEndpoint, this.deployment, this.apiVersion);
    return postJson(url, body, {
      "api-key": this.apiKey
    });
  }

  async completion(model, systemPrompt, userPrompt, jsonSchemaFormat = null) {
    void model;
    const payload = this.#buildPayload(systemPrompt, userPrompt, jsonSchemaFormat);
    const parsed = await this.responsesRaw(payload);
    const text = extractResponseText(parsed);

    if (!text) {
      throw new Error(`No text found in Azure response: ${JSON.stringify(parsed)}`);
    }

    return text;
  }

  #buildPayload(systemPrompt, userPrompt, jsonSchemaFormat = null) {
    const payload = {
      input: [
        {
          role: "system",
          content: [{ type: "input_text", text: systemPrompt }]
        },
        {
          role: "user",
          content: [{ type: "input_text", text: userPrompt }]
        }
      ]
    };

    if (jsonSchemaFormat) {
      payload.text = buildJsonSchemaFormat(jsonSchemaFormat);
    }

    return payload;
  }
}

export { DEFAULT_API_VERSION };
